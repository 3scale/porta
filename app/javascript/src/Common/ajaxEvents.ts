/**
 * Set up AJAX event global listeners for Rails UJS (remote forms and links). They need to be
 * declared with jQuery so that they can be overridden by local listeners (e.stopPropagation).
 * Otherwise they will run after any other listeners, possibly breaking the UX.
 * Example:
 *   document.body.addEventListener('ajax:error', (event) => {
 *     // Do stuff to local component
 *
 *     // Prevent global listener to be called
 *     event.stopPropagation()
 *   })
 *
 * Origin: app/assets/javascripts/ajax_events.js
 */

import $ from 'jquery'

import type { AJAXErrorEvent } from 'Types/rails-ujs'

/**
 * rails-ujs ajax events.
 */
$(document)
  .on('ajax:before', (...args) => {
    console.log('ajax:before on', args)
    window.ThreeScale.showSpinner()
  })
  .on('ajax:beforeSend', (...args) => { console.log('ajax:beforeSend', args) })
  .on('ajax:send', (...args) => { console.log('ajax:send on', args) })
  .on('ajax:stopped', (...args) => { console.log('ajax:stopped', args) })
  .on('ajax:success', (...args) => { console.log('ajax:success', args) })
  .on('ajax:error', (event) => {
    const [response, status] = (event.originalEvent as AJAXErrorEvent<unknown>).detail
    console.error('ajax:error', response, status)
  })
  .on('ajax:complete', (...args) => {
    console.log('ajax:complete', args)
    window.ThreeScale.hideSpinner()
  })

/**
 * window.$.ajax events, triggered by Colorbox.
 */
window.$(document)
  .on('ajaxStart', (...args) => { console.log('ajaxStart', args) })
  .on('ajaxSend', (...args) => { console.log('ajaxSend', args) })
  .on('ajaxSuccess', (...args) => { console.log('ajaxSuccess', args) })
  .on('ajaxComplete', (...args) => { console.log('ajaxComplete', args) })
  .on('ajaxError', (...args) => { console.log('ajaxError', args) })

/**
 * Handle errors in formtastic forms rendered inside a colorbox. The error template is passed a
 * responseText inside the XHR response.
 *
 * DEPRECATED: replace jquery/colorbox modals with Patternfly modals.
 * TODO: add a colorbox specific selector to separate this with non-legacy implementations.
 */
$(document).on('ajax:error', 'form[data-remote]:not(.pf-c-form)', (event) => {
  const [,, xhr] = (event.originalEvent as AJAXErrorEvent<HTMLDocument>).detail

  // If EventTarget is replaceWith'ed immediately, ajax:success and ajax:complete are not called and the spinner will stay
  setTimeout(() => {
    window.$.colorbox({ html: xhr.responseText })
  })
  event.stopPropagation()
})
