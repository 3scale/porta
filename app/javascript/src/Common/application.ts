import type { JQueryXHR } from 'Types/jquery/v1'

const jQuery1 = window.$

export default function (): void {
  // disable links with 'data-disabled' attribute and display alert instead
  // delegation on body fires before rails.js. FIXME: this is not a valid guard. If "data-disabled"
  // removed manually from the DOM, the request will still go through. The link should be disabled
  // (not clickable) and the server should return a flash error.
  $('body').on('click', 'a[data-disabled]', ({ currentTarget }) => {
    alert((currentTarget as HTMLAnchorElement).dataset.disabled)
    return false
  })

  /**
   * This seems to be used only in app/views/master/providers/plans/_widget.html.slim:23
   * DEPRECATED: replace jquery/colorbox modals with Patternfly modals.
   * TODO: add a more specific selector to separate this form from all other "remote" ones.
   */
  $(document).on('submit', 'form.colorbox[data-remote]', ({ currentTarget }) => {
    // MUST USE RAILS-JQUERY since ajax:complete is triggered by rails (do not use $ from node_modules).
    const $form = jQuery1(currentTarget as HTMLFormElement)
    $form.on('ajax:complete', (_event, xhr: JQueryXHR) => {
      jQuery1.colorbox({
        open: true,
        html: xhr.responseText,
        width: $form.data('width') as string | undefined,
        maxWidth: '85%',
        maxHeight: '90%'
      })
    })
  })

  // TODO: replace .fancybox with .colorbox
  // This link will load its content into a colorbox modal
  $(document).on('click', 'a.fancybox, a.colorbox', ({ currentTarget }) => {
    jQuery1(currentTarget as HTMLAnchorElement).colorbox({ open: true })
    return false
  })

  // TODO: replace .fancybox with .colorbox
  // This is used in some modals with a "Cancel" button.
  $(document).on('click', '.fancybox-close', () => {
    jQuery1.colorbox.close()
    return false
  })

  /**
   * Handle errors in formtastic forms rendered inside a colorbox. The error template is passed a
   * responseText inside the XHR response. MUST USE RAILS-JQUERY since ajax:error is triggered by
   * rails (do not use $ from node_modules).
   *
   * DEPRECATED: replace jquery/colorbox modals with Patternfly modals.
   * TODO: add a colorbox specific selector to separate this with non-legacy implementations.
   */
  jQuery1(document).on('ajax:error', 'form:not(.pf-c-form)', (_event, xhr: JQueryXHR, status) => {
    if (status === 'error') {
      jQuery1.colorbox({ html: xhr.responseText })
      return false
    }
  })
}
