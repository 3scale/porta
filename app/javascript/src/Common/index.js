// @flow

export * from './components/SelectModal'
