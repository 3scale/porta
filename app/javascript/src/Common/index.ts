export { CompactListCard } from './components/CompactListCard'
export { FancySelect } from './components/FancySelect'
export { MicroPagination } from './components/MicroPagination'
export { NoMatchFound } from './components/NoMatchFound'
export { TableModal } from './components/TableModal'
export { ToolbarSearch } from './components/ToolbarSearch'
export { Select } from './components/Select'
export { SelectWithModal } from './components/SelectWithModal'
export { Spinner } from './components/Spinner'
export { SystemNamePopover, SystemNamePopoverWrapper } from './components/SystemNamePopover'
export { UserDefinedField } from './components/UserDefinedField'
