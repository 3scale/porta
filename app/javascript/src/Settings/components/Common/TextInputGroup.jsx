// @flow

import * as React from 'react'
import { useState } from 'react'
import { FormGroup, TextInput } from '@patternfly/react-core'
import type { FieldGroupProps } from 'Settings/types'

const TextInputGroup = ({
  defaultValue,
  placeholder,
  label,
  name,
  hint,
  value,
  isDefaultValue = false,
  readOnly = false,
  inputType = 'text'
}: FieldGroupProps): React.Node => {
  const [ inputValue, setInputValue ] = useState(value)
  const onChange = (value, _e) => setInputValue(value)
  return (
    <FormGroup label={label} fieldId={`service_proxy_attributes_${name}_input`} helperText={hint}>
      <TextInput
        id={`service_proxy_attributes_${name}_input`}
        name={`service[proxy_attributes][${name}]`}
        placeholder={placeholder}
        value={isDefaultValue ? defaultValue : inputValue}
        type={inputType}
        onChange={onChange}
        isReadOnly={readOnly}
      />
    </FormGroup>
  )
}

export {
  TextInputGroup
}
