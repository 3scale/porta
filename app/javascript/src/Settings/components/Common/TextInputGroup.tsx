import * as React from 'react'
import { useState } from 'react'

import { FormGroup, TextInput } from '@patternfly/react-core'
import { FieldGroupProps } from 'Settings/types'

const TextInputGroup: React.FunctionComponent<FieldGroupProps> = ({
  defaultValue,
  placeholder,
  label,
  name,
  hint,
  value,
  isDefaultValue = false,
  readOnly = false,
  inputType = 'text'
}) => {
  const [ inputValue, setInputValue ] = useState(value)
  const onChange = (value: string) => setInputValue(value)
  return (
    <FormGroup label={label} fieldId={`service_proxy_attributes_${name}_input`} helperText={hint}>
      <TextInput
        id={`service_proxy_attributes_${name}_input`}
        name={`service[proxy_attributes][${name}]`}
        placeholder={placeholder}
        value={isDefaultValue ? defaultValue : inputValue}
        type={inputType}
        onChange={onChange}
        isReadOnly={readOnly}
      />
    </FormGroup>
  )
}

export { TextInputGroup, FieldGroupProps as Props }
