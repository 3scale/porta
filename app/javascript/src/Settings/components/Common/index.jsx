// @flow

export * from 'Settings/components/Common/FormCollection'
export * from 'Settings/components/Common/FormFieldset'
export * from 'Settings/components/Common/FormLegend'
export * from 'Settings/components/Common/SelectGroup'
export * from 'Settings/components/Common/TextInputGroup'
export * from 'Settings/components/Common/RadioFieldset'
export * from 'Settings/components/Common/TypeItemCombo'
