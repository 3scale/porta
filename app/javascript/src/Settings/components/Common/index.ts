export { FormCollection } from 'Settings/components/Common/FormCollection'
export { FormFieldset } from 'Settings/components/Common/FormFieldset'
export { FormLegend } from 'Settings/components/Common/FormLegend'
export { SelectGroup } from 'Settings/components/Common/SelectGroup'
export { TextInputGroup } from 'Settings/components/Common/TextInputGroup'
export { RadioFieldset } from 'Settings/components/Common/RadioFieldset'
export { TypeItemCombo } from 'Settings/components/Common/TypeItemCombo'
