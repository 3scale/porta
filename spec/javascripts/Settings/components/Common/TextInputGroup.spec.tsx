import { mount } from 'enzyme'
import { TextInputGroup } from 'Settings/components/Common/TextInputGroup'

import type { Props } from 'Settings/components/Common/TextInputGroup'

const defaultProps: Props = {
  name: 'dooz_kawa_chanson',
  value: 'Me Faire La Belle',
  label: '',
  legend: undefined,
  checked: undefined,
  hint: 'Enter your favourite Dooz Kawa tune',
  placeholder: 'You Favourite Dooz Kawa',
  defaultValue: 'Le Monstre',
  readOnly: undefined,
  inputType: undefined,
  isDefaultValue: undefined,
  onChange: undefined
}

const mountWrapper = (props: Partial<Props> = {}) => mount(<TextInputGroup {...{ ...defaultProps, ...props }} />)

it('should render correctly', () => {
  const view = mountWrapper()
  expect(view).toMatchSnapshot()
})

it('should default value when indicated', () => {
  const view = mountWrapper({ isDefaultValue: true })
  expect(view.find('input').prop('value')).toBe('Le Monstre')
})
