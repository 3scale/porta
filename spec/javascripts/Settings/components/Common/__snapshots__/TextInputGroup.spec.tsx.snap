// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<TextInputGroup
  defaultValue="Le Monstre"
  hint="Enter your favourite Dooz Kawa tune"
  label=""
  name="dooz_kawa_chanson"
  placeholder="You Favourite Dooz Kawa"
  value="Me Faire La Belle"
>
  <FormGroup
    fieldId="service_proxy_attributes_dooz_kawa_chanson_input"
    helperText="Enter your favourite Dooz Kawa tune"
    label=""
  >
    <GenerateId
      prefix="pf-random-id-"
    >
      <div
        className="pf-c-form__group"
      >
        <div
          className="pf-c-form__group-control"
        >
          <TextInput
            id="service_proxy_attributes_dooz_kawa_chanson_input"
            isReadOnly={false}
            name="service[proxy_attributes][dooz_kawa_chanson]"
            onChange={[Function]}
            placeholder="You Favourite Dooz Kawa"
            type="text"
            value="Me Faire La Belle"
          >
            <TextInputBase
              aria-label={null}
              className=""
              id="service_proxy_attributes_dooz_kawa_chanson_input"
              innerRef={null}
              isDisabled={false}
              isIconSprite={false}
              isLeftTruncated={false}
              isReadOnly={false}
              isRequired={false}
              name="service[proxy_attributes][dooz_kawa_chanson]"
              onChange={[Function]}
              ouiaSafe={true}
              placeholder="You Favourite Dooz Kawa"
              type="text"
              validated="default"
              value="Me Faire La Belle"
            >
              <input
                aria-invalid={false}
                aria-label={null}
                className="pf-c-form-control"
                data-ouia-component-id="OUIA-Generated-TextInputBase-1"
                data-ouia-component-type="PF4/TextInput"
                data-ouia-safe={true}
                disabled={false}
                id="service_proxy_attributes_dooz_kawa_chanson_input"
                name="service[proxy_attributes][dooz_kawa_chanson]"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="You Favourite Dooz Kawa"
                required={false}
                type="text"
                value="Me Faire La Belle"
              />
            </TextInputBase>
          </TextInput>
          <div
            aria-live="polite"
            className="pf-c-form__helper-text"
            id="service_proxy_attributes_dooz_kawa_chanson_input-helper"
          >
            Enter your favourite Dooz Kawa tune
          </div>
        </div>
      </div>
    </GenerateId>
  </FormGroup>
</TextInputGroup>
`;
